cluster mempool: optimized candidate search

sipa opened the pull request:
Part of cluster mempool: #30289 

Depends on #30126, and was split off from it.

This improves the candidate search algorithm introduced in the previous PR with a variety of optimizations.

The resulting search algorithm largely follows Section 2 of [How to linearize your cluster](https://delvingbitcoin.org/t/how-to-linearize-your-cluster/303#h-2-finding-high-feerate-subsets-5), though with a few changes:
* Connected component analysis is performed inside the search algorithm (creating initial work items per component for each candidate), rather than once at a higher level. This duplicates some work but is significantly simpler in implementation.
* No ancestor-set based presplitting inside the search is performed; instead, the `best` value is initialized with the best topologically valid set known to the LIMO algorithm before search starts: the better one out of the highest-feerate remaining ancestor set, and the highest-feerate prefix of remaining transactions in `old_linearization`.
* Work items are represented using an included set *inc* and an undefined set *und*, rather than included and excluded.
* Potential sets *pot* are not computed for work items with empty *inc*.

At a high level, the only missing optimization from that post is bottleneck analysis; my thinking is that it only really helps with clusters that are already relatively cheap to linearize (doing so would need to be done at a higher level, not inside the search algorithm).

---

Overview of the impact of each commit here on linearize performance:
* **[clusterlin bench: have low/high iter benchmarks instead of per-iter](https://github.com/bitcoin/bitcoin/pull/30286/commits/21a184db63bd13cf0c2dff9e6e61ca85f2ff4454)**: no impact
* **[separate initial search entries per component (optimization)](https://github.com/bitcoin/bitcoin/pull/30286/commits/c84c5c86ba6c12ce11e997d0c59d60694c667488)**: reduce iterations, increase start-up cost
* **[add reordering support for DepGraph](https://github.com/bitcoin/bitcoin/pull/30286/commits/019ff2960976b677d91bab47e75be6ed180b37a3)**: no impact
* **[use feerate-sorted depgraph in SearchCandidateFinder](https://github.com/bitcoin/bitcoin/pull/30286/commits/8e27dd5a220237da5d362376db6406e47f063776)**: typically reduce iterations, increase start-up cost
* **[track upper bound potential set for work items](https://github.com/bitcoin/bitcoin/pull/30286/commits/781e0fb3aa5cbd0983b01547481eefa0804dfb5b)**: reduce iterations, increase cost per iteration
* **[reduce computation of unnecessary pot sets](https://github.com/bitcoin/bitcoin/pull/30286/commits/9fe834fa979c8ab038aec57dac2f468ca70c41a9)**: reduce cost per iteration
* **[include topological pot subsets automatically](https://github.com/bitcoin/bitcoin/pull/30286/commits/30612710a4e5fc8c950bd1ff4ab9fa843b59132d)**: reduce iterations, increase cost per iteration
* **[improve heuristic to decide split transaction](https://github.com/bitcoin/bitcoin/pull/30286/commits/1880c00ab1661dbfc867f69e7556b28bc3cf7a42)**: typically reduce iterations, increase cost per iteration
* **[only start/use search when enough iterations left](https://github.com/bitcoin/bitcoin/pull/30286/commits/12760a57b3a6cd1aeb3b7532311f648de2b45aa4)**: just account for start-up cost as equivalent iterations

sipa committed:
clusterlin: improve heuristic to decide split transaction (optimization)

Empirically, this approach seems to be more efficient in common real-life
clusters, and does not change the worst case.

Co-Authored-By: Suhas Daftuar <sdaftuar@gmail.com>

sipa committed:
clusterlin: include topological pot subsets automatically (optimization)

Automatically add topologically-valid subsets of the potential set pot
to inc. It can be proven that these must be part of the best reachable
topologically-valid set from that work item.

This is a crucial optimization that (apparently) reduces the maximum
number of iterations from ~2^(N-1) to ~sqrt(2^N).

Co-Authored-By: Suhas Daftuar <sdaftuar@gmail.com>

sipa committed:
clusterlin: reduce computation of unnecessary pot sets (optimization)

Keep track of which transactions in the graph have an individual
feerate that is better than the best included set so far. Others do not
need to be added to the pot set, as they cannot possibly help beating
best.

sipa committed:
clusterlin: track upper bound potential set for work items (optimization)

In each work item, keep track of a conservative overestimate of the best
possible feerate that can be reached from it, and then use these to avoid
exploring hopeless work items.

sipa committed:
clusterlin: separate initial search entries per component (optimization)

Before this commit, the worst case for linearization involves clusters which
break apart in several smaller components after the first candidate is
included in the output linearization.

Address this by never considering work items that span multiple components
of what remains of the cluster.

sipa committed:
clusterlin: only start/use search when enough iterations left

glozow commented on src/cluster_linearize.h:
```diff
@@ -711,8 +721,11 @@ class SearchCandidateFinder
-                for (auto pos : und) {
+                // Add entries to pot. We iterate over all undecided transactions whose feerate is
+                // higher than best. While undecided transactions of lower feerate may improve pot
+                // still, if they do, the resulting pot feerate cannot possibly exceed best's (and
```
fbb988a6128b859b204299357cb6bb4133f83e25, are there some extra words?
```suggestion
                // higher than best. While undecided transactions of lower feerate may improve pot,
                // the resulting pot feerate cannot possibly exceed best's (and
```

glozow commented on src/cluster_linearize.h:
```diff
@@ -966,10 +972,18 @@ std::pair<std::vector<ClusterIndex>, bool> Linearize(const DepGraph<SetType>& de
     linearization.reserve(depgraph.TxCount());
     bool optimal = true;
 
+    // Treat the initialization of SearchCandidateFinder as taking N^2/64 (rounded up) iterations.
+    // If we don't have that many, don't start it.
+    uint64_t start_iterations = (uint64_t{depgraph.TxCount()} * depgraph.TxCount() + 63) / 64;
```
Is the idea that, if we have fewer than n^2/64, we can pretty much expect `SearchCandidateFinder` to not make any meaningful progress?

glozow replied:
Ahh I see, thanks

glozow reviewed (commented):
Did an initial read-through

glozow reviewed (commented):
code review ACK. I haven't reviewed the fuzzers yet, doing that next. Tested by writing some benches that run a lot faster with the optimizations.

instagibbs commented on src/cluster_linearize.h:
```diff
@@ -971,10 +977,19 @@ std::pair<std::vector<ClusterIndex>, bool> Linearize(const DepGraph<SetType>& de
     linearization.reserve(depgraph.TxCount());
     bool optimal = true;
 
+    // Treat the initialization of SearchCandidateFinder as taking N^2/64 (rounded up) iterations
```
why 64?

instagibbs commented on src/cluster_linearize.h:
```diff
@@ -987,12 +1002,22 @@ std::pair<std::vector<ClusterIndex>, bool> Linearize(const DepGraph<SetType>& de
-        iterations_left -= iterations_done_now;
+        uint64_t max_iterations_now = 0;
+        if (src_finder) {
+            // Treat the invocation of SearchCandidateFinder::FindCandidateSet() as costing N/4
```
why 4?

instagibbs commented:
reviewed through https://github.com/bitcoin/bitcoin/pull/30286/commits/06ea1016ff40ebcf3e668bdb993a0747904cb40f

instagibbs commented on src/cluster_linearize.h:
```diff
@@ -118,6 +118,28 @@ class DepGraph
+     *
+     * Complexity: O(N^2) where N=depgraph.TxCount().
+     */
+    DepGraph(const DepGraph<SetType>& depgraph, Span<const ClusterIndex> mapping) noexcept : entries(depgraph.TxCount())
```
LGTM

instagibbs commented on src/cluster_linearize.h:
```diff
@@ -723,19 +729,25 @@ class SearchCandidateFinder
+         * - pot: a subset of the "pot" value for the new work item (but a superset of inc).
+         *        It does not need to be the full pot value; missing pot transactions will be added
+         *        to it by add_fn.
+         * - grow_inc: whether to attempt moving transactions from und to inc, if it can be proven
```
I think this has helped a lot, "clusterlin: avoid recomputing potential set on every split (optimization)" now makes sense to me. Lots of interlocking parts, so having small bite-sizes helps.

Can you motivate why `reconsider_pot` is set as true on inclusion of `anc` and not on exclude? 

When `anc` is included, it's included in both `inc` and `pot`, then:

`if (reconsider_pot) consider_inc = pot.transactions - inc.transactions;`

so no part of `anc` will be in `consider_inc`.

glozow replied:
@sipa thanks this is really helpful, particularly this part:

> 4. anc(t) will always include at least one UL transaction (*)
> (*) The jump ahead is exactly what guarantees that anc(t) always includes a UL transaction. If anc(t) was entirely within UH, it would be topologically valid, and thus the parent work item's jump ahead step would have moved it to I already.

However I'm still having trouble with this part:

> Compare this to what happens in an exclusion branch ...
> 4. desc(t) may include a UH transaction, and because of that it is possible that more transactions can be moved from UL to UH.
> 5. Because of the above, it is possible that there are new topologically-valid subsets of I+UH now that can be moved to I. This can only be due to UH entries added in (4).

It seems true that only the entries added in (4) can cause jump-aheads. However, I'm not convinced that existing `pot` transactions cannot be jump-aheadable. For instance, it's possible that we added a new tx that is the parent of a tx that was already in `pot`.

I added an `assert(anc.Overlaps(elem.und - elem.pot.transactions))` in `split_fn`, right before we call `add_fn` for the inclusion branch in order to check that "jump ahead guarantees that anc(t) always includes a UL transaction". The assertion hit, and essentially what I saw was:

1. Going into a split, we had a really high pot feerate, which included tx `C`.
2. For the exclusion branch, desc(t) included lots of high feerate transactions which lowered the pot feerate. `C` is still in pot. At the start of this `add_fn`, we add tx `P` to pot, since its feerate is now high enough.
3. `P` is part of `consider_inc`, and is added to `inc` in the jump-ahead.
4. `C`, which a child of `P`, would be good because anc(C) is a topological subset of `pot`, but is skipped because `reconsider_pot=false`
5. Later, we split on `C`. `C` is part of UH, not UL, which hits the assert. anc(C) = C which is in UH, but it wasn't added to `inc` during a jump-ahead. This is because `C` was added to UH when its ancestor `P` was considered UL, and `reconsider_pot=false` when `P` was added to UH.

Does this make sense? Or maybe my `assert` isn't the right interpretation of this? I don't think this changes what `FindCandidateSet` outputs, but it means I still don't quite understand the " avoid jump ahead in unmodified pot sets" commit :(

(fwiw the fuzz I crashed on is `echo "AQBJCAgAIysjIyMiIyMjIyMjIyMjIyM6IyMIACMrIyMjIiMjIyMjIyMjIyMjIyMjCBlE+xEvRG9v
AggAbwAR" | base 64 -d`)

glozow replied:
How significant do we expect the `reconsider_pot` optimization to be? Locally my `LinearizePerIter*TxWorstCase` bench doesn't seem to do better after c68d26a0e97.

glozow replied:
Here's what I get on 255571339e7:
```
|               ns/op |                op/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|          724,714.00 |            1,379.85 |    2.4% |      0.01 | `Linearize32TxWorstCase15000Iters`
|          245,712.75 |            4,069.79 |    0.3% |      0.01 | `Linearize32TxWorstCase5000Iters`
|          688,589.50 |            1,452.24 |    5.2% |      0.01 | :wavy_dash: `Linearize48TxWorstCase15000Iters` (Unstable with ~1.2 iters. Increase `minEpochIterations` to e.g. 12)
|          233,424.25 |            4,284.05 |    4.8% |      0.01 | `Linearize48TxWorstCase5000Iters`
|          662,959.00 |            1,508.39 |    1.7% |      0.01 | `Linearize64TxWorstCase15000Iters`
|          242,351.75 |            4,126.23 |    1.9% |      0.01 | `Linearize64TxWorstCase5000Iters`
|        1,051,925.00 |              950.64 |    4.3% |      0.01 | `Linearize75TxWorstCase15000Iters`
|          344,038.00 |            2,906.66 |    1.8% |      0.01 | `Linearize75TxWorstCase5000Iters`
|        1,023,533.00 |              977.01 |    2.1% |      0.01 | `Linearize99TxWorstCase15000Iters`
|          356,557.67 |            2,804.60 |    1.3% |      0.01 | `Linearize99TxWorstCase5000Iters`
```
on d04997871f5:
```
|               ns/op |                op/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|          705,210.00 |            1,418.02 |    4.4% |      0.01 | `Linearize32TxWorstCase15000Iters`
|          231,874.00 |            4,312.69 |    0.2% |      0.01 | `Linearize32TxWorstCase5000Iters`
|          664,206.00 |            1,505.56 |    4.7% |      0.01 | `Linearize48TxWorstCase15000Iters`
|          222,808.25 |            4,488.16 |    1.9% |      0.01 | `Linearize48TxWorstCase5000Iters`
|          627,877.00 |            1,592.67 |    0.2% |      0.01 | `Linearize64TxWorstCase15000Iters`
|          223,255.50 |            4,479.17 |    2.6% |      0.01 | `Linearize64TxWorstCase5000Iters`
|          993,733.00 |            1,006.31 |    3.8% |      0.01 | `Linearize75TxWorstCase15000Iters`
|          327,429.67 |            3,054.09 |    2.3% |      0.01 | `Linearize75TxWorstCase5000Iters`
|          957,057.00 |            1,044.87 |    3.9% |      0.01 | `Linearize99TxWorstCase15000Iters`
|          339,669.33 |            2,944.04 |    1.8% |      0.01 | `Linearize99TxWorstCase5000Iters`
```

Looks somewhat better but not much / close to the err%:

| Benchmark                            | Table1 ns/op | Table2 ns/op | Percentage Speedup |
|--------------------------------------|--------------|--------------|--------------------|
| `Linearize32TxWorstCase15000Iters`   | 724,714.00   | 705,210.00   | 2.70%              |
| `Linearize32TxWorstCase5000Iters`    | 245,712.75   | 231,874.00   | 5.64%              |
| `Linearize48TxWorstCase15000Iters`   | 688,589.50   | 664,206.00   | 3.54%              |
| `Linearize48TxWorstCase5000Iters`    | 233,424.25   | 222,808.25   | 4.55%              |
| `Linearize64TxWorstCase15000Iters`   | 662,959.00   | 627,877.00   | 5.30%              |
| `Linearize64TxWorstCase5000Iters`    | 242,351.75   | 223,255.50   | 7.88%              |
| `Linearize75TxWorstCase15000Iters`   | 1,051,925.00 | 993,733.00   | 5.54%              |
| `Linearize75TxWorstCase5000Iters`    | 344,038.00   | 327,429.67   | 4.84%              |
| `Linearize99TxWorstCase15000Iters`   | 1,023,533.00 | 957,057.00   | 6.49%              |
| `Linearize99TxWorstCase5000Iters`    | 356,557.67   | 339,669.33   | 4.74%              |

sipa committed:
clusterlin: add reordering support for DepGraph

Add a DepGraph(depgraph, reordering) function that constructs a new DepGraph
corresponding to an old one, but with its transactions is a modified order
(given as a vector from old to new positions).

Also use this reordering feature inside DepGraphFormatter::Unser, which needs
a small modification so that its reordering mapping is old-to-new (rather than
the new-to-old it used before).

sipa committed:
clusterlin: use feerate-sorted depgraph in SearchCandidateFinder

This is a requirement for a future commit, which will rely on quickly iterating
over transaction sets in decreasing individual feerate order.

glozow commented on src/cluster_linearize.h:
```diff
@@ -596,56 +678,147 @@ class SearchCandidateFinder
+                    // individual feerate order.
+                    if (!(m_depgraph.FeeRate(pos) >> pot.feerate)) break;
+                    pot.Set(m_depgraph, pos);
+                    consider_inc.Set(pos);
```
Oh of course, I was confusing myself :+1:

sipa committed:
clusterlin bench: have low/high iter benchmarks instead of per-iter

glozow commented on src/bench/cluster_linearize.cpp:
```diff
@@ -203,12 +202,18 @@ void BenchMergeLinearizationsWorstCase(ClusterIndex ntx, benchmark::Bench& bench
-static void LinearizePerIter75TxWorstCase(benchmark::Bench& bench) { BenchLinearizePerIterWorstCase<BitSet<75>>(75, bench); }
-static void LinearizePerIter99TxWorstCase(benchmark::Bench& bench) { BenchLinearizePerIterWorstCase<BitSet<99>>(99, bench); }
+static void Linearize16TxWorstCase50Iters(benchmark::Bench& bench) { BenchLinearizeWorstCase<BitSet<16>>(16, bench, 50); }
+static void Linearize16TxWorstCase150Iters(benchmark::Bench& bench) { BenchLinearizeWorstCase<BitSet<16>>(16, bench, 150); }
```
CI is failing because it halts at 128 iters, it seems.
Original iters given was 2^(16/2-1) = 128.
However, sqrt(2^(n-1)) = sqrt(2^15) > 150

instagibbs reviewed (approved):
code review ACK 1b8f143fb791fe5ba40ce6610cd4e8e04b951495

Thanks for dropping those couple commits.

> I think this will need analysis with real-world clusters to see actual impact, which may be easier once the cluster mempool project is further along. I've dropped it for this reason.

On that note I think we'll need some diagram-quality style benchmarks to make informed decisions in the future, otherwise it makes ongoing changes difficult to justify.

glozow reviewed (commented):
ACK 12760a57b3a

Agree with keeping "reduce computation of unnecessary pot sets", intuitively it seems useful in the average situation and quite cheap.

Thanks for dropping "avoid jump ahead in unmodified pot sets," I wasn't convinced that processing the existing pot txns in the exclusion branch is always redundant or faster to skip. It also made the code harder for future people to understand, which I think is important.

I redid code review, all pretty straightforward now.

Prettified bench results:
![image](https://github.com/user-attachments/assets/a1e45724-cc1e-446a-b346-888983f15639)

[bench.csv](https://github.com/user-attachments/files/16729510/2024-08-22-bench-30286.csv)

instagibbs commented on src/test/fuzz/cluster_linearize.cpp:
```diff
@@ -513,9 +548,11 @@ FUZZ_TARGET(clusterlin_search_finder)
-        // At most 2^N-1 iterations can be required: the number of non-empty subsets a graph with N
-        // transactions has.
-        assert(iterations_done <= ((uint64_t{1} << todo.Count()) - 1));
+        // At most 2^(N-1) iterations can be required: the maximum number of non-empty topological
+        // subsets a (connected) cluster with N transactions can have. Even when the cluster is no
+        // longer connected after removing certain transactions, this holds, because the connected
+        // components are searched separately.
```
(in case my reasoning was faulty) My example was N-1 parents and 1 child.

2^(N-1) - 1 possible choices of parents without child
Adding child is one more possibility with all parents selected, resulting in 2^(N-1)?

instagibbs replied:
@sipa I psyop'd myself into thinking they were connected... somehow.. Well, I believe your example!

ismaelsadeeq reviewed (commented):
Concept ACK

ismaelsadeeq reviewed (commented):
I was hoping to see the benchmark improve but it seems it doesn't but it decreases a bit after the optimizations of this PR.

Benched on M2 Pro

On 21a184db63bd13cf0c2dff9e6e61ca85f2ff4454
|               ns/op |                op/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|           34,200.59 |           29,239.26 |    0.4% |      0.01 | `Linearize16TxWorstCase120Iters`
|            6,384.70 |          156,624.46 |    2.1% |      0.01 | `Linearize16TxWorstCase20Iters`
|        3,416,000.00 |              292.74 |    1.0% |      0.04 | `Linearize32TxWorstCase15000Iters`
|        1,159,875.00 |              862.16 |    1.7% |      0.01 | `Linearize32TxWorstCase5000Iters`
|        3,372,792.00 |              296.49 |    2.1% |      0.04 | `Linearize48TxWorstCase15000Iters`
|        1,144,166.00 |              874.00 |    2.2% |      0.01 | `Linearize48TxWorstCase5000Iters`
|        3,365,333.00 |              297.15 |    1.8% |      0.04 | `Linearize64TxWorstCase15000Iters`
|        1,122,792.00 |              890.64 |    0.7% |      0.01 | `Linearize64TxWorstCase5000Iters`
|        5,721,541.00 |              174.78 |    1.0% |      0.06 | `Linearize75TxWorstCase15000Iters`
|        1,904,166.00 |              525.16 |    1.1% |      0.02 | `Linearize75TxWorstCase5000Iters`
|        5,672,875.00 |              176.28 |    1.3% |      0.06 | `Linearize99TxWorstCase15000Iters`
|        1,889,542.00 |              529.23 |    0.8% |      0.02 | `Linearize99TxWorstCase5000Iters`


On 12760a57b3a6cd1aeb3b7532311f648de2b45aa4

|               ns/op |                op/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|          152,479.17 |            6,558.27 |    1.3% |      0.01 | `Linearize16TxWorstCase120Iters`
|           40,915.00 |           24,440.91 |    2.3% |      0.01 | `Linearize16TxWorstCase20Iters`
|       12,573,458.00 |               79.53 |    1.3% |      0.14 | `Linearize32TxWorstCase15000Iters`
|        4,422,792.00 |              226.10 |    1.0% |      0.05 | `Linearize32TxWorstCase5000Iters`
|       12,813,167.00 |               78.04 |    0.5% |      0.14 | `Linearize48TxWorstCase15000Iters`
|        4,635,208.00 |              215.74 |    1.1% |      0.05 | `Linearize48TxWorstCase5000Iters`
|       13,188,833.00 |               75.82 |    1.2% |      0.14 | `Linearize64TxWorstCase15000Iters`
|        4,867,542.00 |              205.44 |    1.0% |      0.05 | `Linearize64TxWorstCase5000Iters`
|       20,246,083.00 |               49.39 |    1.3% |      0.22 | `Linearize75TxWorstCase15000Iters`
|        7,692,334.00 |              130.00 |    1.1% |      0.08 | `Linearize75TxWorstCase5000Iters`
|       20,629,500.00 |               48.47 |    2.2% |      0.22 | `Linearize99TxWorstCase15000Iters`
|        7,990,417.00 |              125.15 |    1.7% |      0.09 | `Linearize99TxWorstCase5000Iters`

I believe this is due to the increased per-iteration cost introduced by the optimizations. In favor of reduction of the maximum number of iterations from  $2^{N-1}$ to $\sqrt{2^N}$ (which is really nice).



One notable thing I observe from this result is as the total number of iterations increase the total time taken also increases a bit.

For smaller transaction clusters which are more common in the mempool is the tradeoff is justified given that the total time to complete in master and this PR is fairly even from the bench?



---
#### I attempted to benchmark the maximum `iter_limit` values before and after this PR on few cluster sizes

I did the benchmark by modifying the `iter_limit` parameter of the `BenchLinearizeWorstCase` function for various cluster sizes (16, 32, 48, 64, 75, and 99).


On [21a184db63bd13cf0c2dff9e6e61ca85f2ff4454](https://github.com/bitcoin/bitcoin/commit/21a184db63bd13cf0c2dff9e6e61ca85f2ff4454)

I set  `iter_limit` to $2^{N-1}$

<details>
<summary>View Diff</summary>

```diff
diff --git a/src/bench/cluster_linearize.cpp b/src/bench/cluster_linearize.cpp
index e63e5d76c8b..d18df32b202 100644
--- a/src/bench/cluster_linearize.cpp
+++ b/src/bench/cluster_linearize.cpp
@@ -114,14 +114,17 @@ DepGraph<SetType> MakeHardGraph(ClusterIndex ntx)
  * iterations difference.
  */
 template<typename SetType>
-void BenchLinearizeWorstCase(ClusterIndex ntx, benchmark::Bench& bench, uint64_t iter_limit)
+void BenchLinearizeWorstCase(ClusterIndex ntx, benchmark::Bench& bench, std::optional<uint64_t> iter_limit)
 {
+    if (!iter_limit.has_value()) {
+        iter_limit = 1 << (ntx - 1);
+    }
     const auto depgraph = MakeHardGraph<SetType>(ntx);
     uint64_t rng_seed = 0;
     bench.run([&] {
         SearchCandidateFinder finder(depgraph, rng_seed++);
-        auto [candidate, iters_performed] = finder.FindCandidateSet(iter_limit, {});
-        assert(iters_performed == iter_limit);
+        auto [candidate, iters_performed] = finder.FindCandidateSet(iter_limit.value(), {});
+        assert(iters_performed <= iter_limit.value());
     });
 }
 
@@ -217,6 +220,13 @@ static void Linearize75TxWorstCase15000Iters(benchmark::Bench& bench) { BenchLin
 static void Linearize99TxWorstCase5000Iters(benchmark::Bench& bench) { BenchLinearizeWorstCase<BitSet<99>>(99, bench, 5000); }
 static void Linearize99TxWorstCase15000Iters(benchmark::Bench& bench) { BenchLinearizeWorstCase<BitSet<99>>(99, bench, 15000); }

+static void Linearize16TxWorstCase(benchmark::Bench& bench) { BenchLinearizeWorstCase<BitSet<16>>(16, bench, std::nullopt); }
+static void Linearize32TxWorstCase(benchmark::Bench& bench) { BenchLinearizeWorstCase<BitSet<32>>(32, bench, std::nullopt); }
+static void Linearize48TxWorstCase(benchmark::Bench& bench) { BenchLinearizeWorstCase<BitSet<48>>(48, bench, std::nullopt); }
+static void Linearize64TxWorstCase(benchmark::Bench& bench) { BenchLinearizeWorstCase<BitSet<64>>(64, bench, std::nullopt); }
+static void Linearize75TxWorstCase(benchmark::Bench& bench) { BenchLinearizeWorstCase<BitSet<75>>(75, bench, std::nullopt); }
+static void Linearize99TxWorstCase(benchmark::Bench& bench) { BenchLinearizeWorstCase<BitSet<99>>(99, bench, std::nullopt); }
+
 static void LinearizeNoIters16TxWorstCaseAnc(benchmark::Bench& bench) { BenchLinearizeNoItersWorstCaseAnc<BitSet<16>>(16, bench); }
 static void LinearizeNoIters32TxWorstCaseAnc(benchmark::Bench& bench) { BenchLinearizeNoItersWorstCaseAnc<BitSet<32>>(32, bench); }
 static void LinearizeNoIters48TxWorstCaseAnc(benchmark::Bench& bench) { BenchLinearizeNoItersWorstCaseAnc<BitSet<48>>(48, bench); }
@@ -245,43 +255,50 @@ static void MergeLinearizations64TxWorstCase(benchmark::Bench& bench) { BenchMer
 static void MergeLinearizations75TxWorstCase(benchmark::Bench& bench) { BenchMergeLinearizationsWorstCase<BitSet<75>>(75, bench); }
 static void MergeLinearizations99TxWorstCase(benchmark::Bench& bench) { BenchMergeLinearizationsWorstCase<BitSet<99>>(99, bench); }

+BENCHMARK(Linearize16TxWorstCase, benchmark::PriorityLevel::HIGH);
+BENCHMARK(Linearize32TxWorstCase, benchmark::PriorityLevel::HIGH);
+BENCHMARK(Linearize48TxWorstCase, benchmark::PriorityLevel::HIGH);
+BENCHMARK(Linearize64TxWorstCase, benchmark::PriorityLevel::HIGH);
+BENCHMARK(Linearize75TxWorstCase, benchmark::PriorityLevel::HIGH);
+BENCHMARK(Linearize99TxWorstCase, benchmark::PriorityLevel::HIGH);

```

</details>


|           ns/op |         op/s | err% | total | Benchmark                    |
|----------------:|-------------:|-----:|------:|:-----------------------------|
|   1,168,959.00  |      855.46  | 1.8% |  0.01 | `Linearize16TxWorstCase`     |
| 5,546,978,667.00|        0.18  | 0.8% | 60.83 | `Linearize32TxWorstCase`     |
|   7,401,916.00  |      135.10  | 0.4% |  0.08 | `Linearize48TxWorstCase`     |
|     421,208.50  |    2,374.12  | 0.8% |  0.01 | `Linearize75TxWorstCase`     |
|       4,004.02  |  249,749.05  | 0.6% |  0.01 | `Linearize99TxWorstCase`     |

The benchmark for `Linearize64TxWorstCase` hung during execution, while the others completed successfully.
`Linearize32TxWorstCase` completed after a minute.




On [12760a57b3a6cd1aeb3b7532311f648de2b45aa4](https://github.com/bitcoin/bitcoin/commit/12760a57b3a6cd1aeb3b7532311f648de2b45aa4)

I set `iter_limit`` to $√{2^{N}}$



<details>
<summary>View Diff</summary>

```diff
diff --git a/src/bench/cluster_linearize.cpp b/src/bench/cluster_linearize.cpp
index 48ede53bd7d..e50bc55f4d9 100644
--- a/src/bench/cluster_linearize.cpp
+++ b/src/bench/cluster_linearize.cpp
@@ -6,6 +6,7 @@
 
 #include <util/bitset.h>
 #include <cluster_linearize.h>
+#include <cmath>
 
 using namespace cluster_linearize;
 
@@ -114,14 +115,17 @@ DepGraph<SetType> MakeHardGraph(ClusterIndex ntx)
  * iterations difference.
  */
 template<typename SetType>
-void BenchLinearizeWorstCase(ClusterIndex ntx, benchmark::Bench& bench, uint64_t iter_limit)
+void BenchLinearizeWorstCase(ClusterIndex ntx, benchmark::Bench& bench, std::optional<uint64_t> iter_limit)
 {
+    if (!iter_limit.has_value()) {
+        iter_limit = std::sqrt(1 << ntx);
+    }
     const auto depgraph = MakeHardGraph<SetType>(ntx);
     uint64_t rng_seed = 0;
     bench.run([&] {
         SearchCandidateFinder finder(depgraph, rng_seed++);
-        auto [candidate, iters_performed] = finder.FindCandidateSet(iter_limit, {});
-        assert(iters_performed == iter_limit);
+        auto [candidate, iters_performed] = finder.FindCandidateSet(iter_limit.value(), {});
+        assert(iters_performed <= iter_limit.value());
     });
 }
 
@@ -217,6 +221,13 @@ static void Linearize75TxWorstCase15000Iters(benchmark::Bench& bench) { BenchLin
 static void Linearize99TxWorstCase5000Iters(benchmark::Bench& bench) { BenchLinearizeWorstCase<BitSet<99>>(99, bench, 5000); }
 static void Linearize99TxWorstCase15000Iters(benchmark::Bench& bench) { BenchLinearizeWorstCase<BitSet<99>>(99, bench, 15000); }

+static void Linearize16TxWorstCase(benchmark::Bench& bench) { BenchLinearizeWorstCase<BitSet<16>>(16, bench, std::nullopt); }
+static void Linearize32TxWorstCase(benchmark::Bench& bench) { BenchLinearizeWorstCase<BitSet<32>>(32, bench, std::nullopt); }
+static void Linearize48TxWorstCase(benchmark::Bench& bench) { BenchLinearizeWorstCase<BitSet<48>>(48, bench, std::nullopt); }
+static void Linearize64TxWorstCase(benchmark::Bench& bench) { BenchLinearizeWorstCase<BitSet<64>>(64, bench, std::nullopt); }
+static void Linearize75TxWorstCase(benchmark::Bench& bench) { BenchLinearizeWorstCase<BitSet<75>>(75, bench, std::nullopt); }
+static void Linearize99TxWorstCase(benchmark::Bench& bench) { BenchLinearizeWorstCase<BitSet<99>>(99, bench, std::nullopt); }
+
 static void LinearizeNoIters16TxWorstCaseAnc(benchmark::Bench& bench) { BenchLinearizeNoItersWorstCaseAnc<BitSet<16>>(16, bench); }
 static void LinearizeNoIters32TxWorstCaseAnc(benchmark::Bench& bench) { BenchLinearizeNoItersWorstCaseAnc<BitSet<32>>(32, bench); }
 static void LinearizeNoIters48TxWorstCaseAnc(benchmark::Bench& bench) { BenchLinearizeNoItersWorstCaseAnc<BitSet<48>>(48, bench); }
@@ -245,43 +256,50 @@ static void MergeLinearizations64TxWorstCase(benchmark::Bench& bench) { BenchMer
 static void MergeLinearizations75TxWorstCase(benchmark::Bench& bench) { BenchMergeLinearizationsWorstCase<BitSet<75>>(75, bench); }
 static void MergeLinearizations99TxWorstCase(benchmark::Bench& bench) { BenchMergeLinearizationsWorstCase<BitSet<99>>(99, bench); }

+BENCHMARK(Linearize16TxWorstCase, benchmark::PriorityLevel::HIGH);
+BENCHMARK(Linearize32TxWorstCase, benchmark::PriorityLevel::HIGH);
+BENCHMARK(Linearize48TxWorstCase, benchmark::PriorityLevel::HIGH);
+BENCHMARK(Linearize64TxWorstCase, benchmark::PriorityLevel::HIGH);
+BENCHMARK(Linearize75TxWorstCase, benchmark::PriorityLevel::HIGH);
+BENCHMARK(Linearize99TxWorstCase, benchmark::PriorityLevel::HIGH);


```

</details>


|         ns/op |        op/s | err% | total | Benchmark                |
|--------------:|------------:|-----:|------:|:-------------------------|
|   168,569.50  |   5,932.27  | 0.6% |  0.01 | `Linearize16TxWorstCase` |
|    18,648.34  |  53,624.08  | 0.3% |  0.01 | `Linearize32TxWorstCase` |
|   770,417.00  |   1,298.00  | 1.0% |  0.01 | `Linearize48TxWorstCase` |
|    34,281.25  |  29,170.46  | 0.3% |  0.01 | `Linearize64TxWorstCase` |
|   379,319.67  |   2,636.30  | 2.2% |  0.01 | `Linearize75TxWorstCase` |
|    78,330.08  |  12,766.49  | 2.2% |  0.01 | `Linearize99TxWorstCase` |

All benchmarks completed successfully, and the results indicate a performance improvement compared to the previous test.

- Setting `iter_limit` to $2^{N-1}$ before this PR resulted in longer execution times and, in one case (`Linearize64TxWorstCase`), caused Linearize64TxWorstCase benchmark to hang which I did not investigate why.

- Setting the `iter_limit` to $\sqrt{2^{N}}$ improved performance across all cluster sizes, with all benchmarks completing successfully.

ismaelsadeeq commented on src/cluster_linearize.h:
```diff
@@ -668,7 +668,9 @@ class SearchCandidateFinder
-             *  conservative upper bound on how good a set this work item can give rise to. */
+             *  conservative upper bound on how good a set this work item can give rise to. If the
+             *  real best such feerate does not exceed best's, then this value is not guaranteed to
+             *  be accurate. */
```
The new comment is much better, thanks for the example here.

ismaelsadeeq reviewed (commented):
light code review and Tested ACK 100e1b9ecb29c76187112fda9fed1c01da192c99

glozow commented:
reACK 100e1b9ecb2 via range-diff

sdaftuar commented:
> @sdaftuar do you think you could instrument your code to gather/dump clusters that linearize optimally with say 1,000,000 iterations but not using 100,000 iterations, using the code up to commit [21b826a](https://github.com/bitcoin/bitcoin/commit/21b826af07fdfa26aaf981cc0810edd6b55b45e0) (before that commit the search algorithm is so naive that high iteration count isn't really an indication of being hard).

I used my simulation environment to run the commit you listed on last year's data, searching for clusters that were optimal within 1M iterations but required more than 100,000 iterations.  I then sorted those by iteration count and took the "worst" 20 clusters, which I added to our benchmark suite in this branch: https://github.com/sdaftuar/bitcoin/tree/test-30286

I then ran that benchmark on each intermediate commit of this PR.  At the beginning of this PR, on my machine, the fastest cluster takes **~5ms** to linearize, with nearly all the rest being at around 10ms.  By the end of this PR, the fastest cluster linearizes (optimally) in **1.4us**, and the slowest cluster takes **115us** (and other than the slowest cluster I found, all the rest are under **50us**).  So this is a huge improvement.

Note that I excluded any LIMO calculations in this benchmark (by not passing in a pre-existing linearization to `Linearize`).

I'm attaching my output here for others to review/double check my summary (but you can also try to recreate the benchmarks locally using the branch I shared above): 
[output.bench.txt](https://github.com/user-attachments/files/16871148/output.bench.txt)

instagibbs commented:
@sdaftuar awesome data! Encouraging to see the timings move along with what @sipa was asserting :+1:

sipa committed:
clusterlin bench: add example hard cluster benchmarks

Co-Authored-By: Suhas Daftuar <sdaftuar@gmail.com>

instagibbs commented on src/bench/cluster_linearize.cpp:
```diff
@@ -207,6 +210,42 @@ void BenchMergeLinearizationsWorstCase(ClusterIndex ntx, benchmark::Bench& bench
+        reader >> Using<DepGraphFormatter>(depgraph);
+        uint64_t rng_seed = 0;
+        bench.run([&] {
+            auto res = Linearize(depgraph, /*max_iterations=*/1000000, rng_seed++);
```
is f94c546ac02ad11ba73174cb8d25220515cee893 the first commit that you can assert optimality? I'm not sure having the benchmark in prior to that is so meaningful

instagibbs replied:
also seems like the assertion fails for me at that commit

glozow commented:
Examples 16-19 seem to get slower?
My results on the last commit look similar to sipa's:

|               ns/op |                op/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|       17,187,614.00 |               58.18 |    4.0% |      0.19 | `LinearizeOptimallyExample00`
|           96,365.70 |           10,377.14 |    0.2% |      0.01 | `LinearizeOptimallyExample01`
|           87,832.50 |           11,385.31 |    3.1% |      0.01 | `LinearizeOptimallyExample02`
|           86,292.45 |           11,588.50 |    1.9% |      0.01 | `LinearizeOptimallyExample03`
|            8,905.69 |          112,287.78 |    1.3% |      0.01 | `LinearizeOptimallyExample04`
|           90,947.18 |           10,995.39 |    3.2% |      0.01 | `LinearizeOptimallyExample05`
|           90,872.91 |           11,004.38 |    4.2% |      0.01 | `LinearizeOptimallyExample06`
|           93,178.40 |           10,732.10 |    1.3% |      0.01 | `LinearizeOptimallyExample07`
|           72,569.92 |           13,779.81 |    3.9% |      0.01 | `LinearizeOptimallyExample08`
|          106,545.57 |            9,385.66 |    3.8% |      0.01 | `LinearizeOptimallyExample09`
|          103,444.11 |            9,667.06 |    0.9% |      0.01 | `LinearizeOptimallyExample10`
|           74,771.30 |           13,374.12 |    2.9% |      0.01 | `LinearizeOptimallyExample11`
|           57,063.53 |           17,524.33 |    0.3% |      0.01 | `LinearizeOptimallyExample12`
|           43,317.04 |           23,085.60 |    2.1% |      0.01 | `LinearizeOptimallyExample13`
|           82,872.00 |           12,066.80 |    2.6% |      0.01 | `LinearizeOptimallyExample14`
|           21,907.07 |           45,647.37 |    0.8% |      0.01 | `LinearizeOptimallyExample15`
|      276,053,164.00 |                3.62 |    0.3% |      3.03 | `LinearizeOptimallyExample16`
|      275,137,154.00 |                3.63 |    0.2% |      3.04 | `LinearizeOptimallyExample17`
|      258,526,924.00 |                3.87 |    0.5% |      2.85 | `LinearizeOptimallyExample18`
|      273,558,093.00 |                3.66 |    0.3% |      3.02 | `LinearizeOptimallyExample19`

But here's commit-by-commit [2024-09-06-bench-examples.txt](https://github.com/user-attachments/files/16908467/2024-09-06-bench-examples.txt)
![image](https://github.com/user-attachments/assets/43e6ed6a-47ba-4b89-8582-76499d80a257)

sdaftuar commented:
> @sdaftuar Is it possible that some of these clusters are "related" (as in are really modifications of one another with a few transactions added/removed)? That could explain the strong correlations between consecutive ones.

I do believe those last 4 examples are likely to be related, as all were found within minutes of each other on the same day.

instagibbs commented:
Confirmed that at least for example 19, it simply can't find the optimal solution until the commit it's included, even if total iterations are cranked up a couple orders of magnitude.

instagibbs commented:
reACK 59028de422bbf8ccf53da27f1153e343952e585f

via `git range-diff master 100e1b9ecb29c76187112fda9fed1c01da192c99 59028de422bbf8ccf53da27f1153e343952e585f`

fwiw I ran the branch with the two dropped optimizations, and it lead to a slight improvement to most of the optimal benchmarks on my machine

glozow commented:
ACK 59028de422b

ismaelsadeeq reviewed (commented):
Re-ACK 59028de422bbf8ccf53da27f1153e343952e585f

I am having consistently similar result like @glozow although some bench were unstable.
<details><summary>bench result</summary>

|               ns/op |                op/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|           13,409.81 |           74,572.28 |    8.4% |      0.01 | :wavy_dash: `Linearize16TxWorstCase120Iters` (Unstable with ~71.1 iters. Increase `minEpochIterations` to e.g. 711)
|            2,857.30 |          349,980.23 |    1.1% |      0.01 | `Linearize16TxWorstCase20Iters`
|          824,833.00 |            1,212.37 |    2.1% |      0.01 | `Linearize32TxWorstCase15000Iters`
|          263,687.50 |            3,792.37 |    3.1% |      0.01 | `Linearize32TxWorstCase5000Iters`
|          685,583.00 |            1,458.61 |    4.4% |      0.01 | `Linearize48TxWorstCase15000Iters`
|          217,283.20 |            4,602.29 |    1.3% |      0.01 | `Linearize48TxWorstCase5000Iters`
|          566,271.00 |            1,765.94 |    5.2% |      0.01 | :wavy_dash: `Linearize64TxWorstCase15000Iters` (Unstable with ~1.7 iters. Increase `minEpochIterations` to e.g. 17)
|          207,125.00 |            4,828.00 |    5.2% |      0.01 | :wavy_dash: `Linearize64TxWorstCase5000Iters` (Unstable with ~4.9 iters. Increase `minEpochIterations` to e.g. 49)
|          794,375.00 |            1,258.85 |    2.2% |      0.01 | `Linearize75TxWorstCase15000Iters`
|          287,364.50 |            3,479.90 |    2.8% |      0.01 | `Linearize75TxWorstCase5000Iters`
|          774,791.00 |            1,290.67 |    3.1% |      0.01 | `Linearize99TxWorstCase15000Iters`
|          286,264.00 |            3,493.28 |    1.6% |      0.01 | `Linearize99TxWorstCase5000Iters`
|              711.44 |        1,405,591.29 |    2.0% |      0.01 | `LinearizeNoIters16TxWorstCaseAnc`
|              811.92 |        1,231,654.51 |    1.1% |      0.01 | `LinearizeNoIters16TxWorstCaseLIMO`
|            2,209.56 |          452,579.03 |    2.0% |      0.01 | `LinearizeNoIters32TxWorstCaseAnc`
|            2,791.17 |          358,272.43 |    2.1% |      0.01 | `LinearizeNoIters32TxWorstCaseLIMO`
|            4,802.87 |          208,208.65 |    2.0% |      0.01 | `LinearizeNoIters48TxWorstCaseAnc`
|            6,225.00 |          160,642.57 |    1.7% |      0.01 | `LinearizeNoIters48TxWorstCaseLIMO`
|            8,375.67 |          119,393.48 |    2.2% |      0.01 | `LinearizeNoIters64TxWorstCaseAnc`
|           11,179.56 |           89,448.96 |    1.9% |      0.01 | `LinearizeNoIters64TxWorstCaseLIMO`
|           13,085.68 |           76,419.44 |    2.9% |      0.01 | `LinearizeNoIters75TxWorstCaseAnc`
|           20,939.10 |           47,757.55 |    1.8% |      0.01 | `LinearizeNoIters75TxWorstCaseLIMO`
|           22,215.91 |           45,012.79 |    0.4% |      0.01 | `LinearizeNoIters99TxWorstCaseAnc`
|           37,518.69 |           26,653.38 |    0.8% |      0.01 | `LinearizeNoIters99TxWorstCaseLIMO`
|       12,650,708.00 |               79.05 |    5.8% |      0.14 | :wavy_dash: `LinearizeOptimallyExample00` (Unstable with ~1.0 iters. Increase `minEpochIterations` to e.g. 10)
|           70,053.57 |           14,274.79 |    1.9% |      0.01 | `LinearizeOptimallyExample01`
|           63,852.80 |           15,661.02 |    1.6% |      0.01 | `LinearizeOptimallyExample02`
|           63,375.00 |           15,779.09 |    1.8% |      0.01 | `LinearizeOptimallyExample03`
|            7,607.93 |          131,441.79 |    1.0% |      0.01 | `LinearizeOptimallyExample04`
|           78,541.60 |           12,732.11 |    6.1% |      0.01 | :wavy_dash: `LinearizeOptimallyExample05` (Unstable with ~7.0 iters. Increase `minEpochIterations` to e.g. 70)
|           68,958.33 |           14,501.51 |    2.5% |      0.01 | `LinearizeOptimallyExample06`
|           75,253.73 |           13,288.38 |    7.2% |      0.01 | :wavy_dash: `LinearizeOptimallyExample07` (Unstable with ~12.4 iters. Increase `minEpochIterations` to e.g. 124)
|           55,020.90 |           18,174.91 |    9.7% |      0.01 | :wavy_dash: `LinearizeOptimallyExample08` (Unstable with ~12.5 iters. Increase `minEpochIterations` to e.g. 125)
|           77,760.42 |           12,860.01 |    3.2% |      0.01 | `LinearizeOptimallyExample09`
|           76,808.30 |           13,019.43 |    1.6% |      0.01 | `LinearizeOptimallyExample10`
|           53,569.44 |           18,667.36 |    2.1% |      0.01 | `LinearizeOptimallyExample11`
|           43,802.08 |           22,829.96 |    0.8% |      0.01 | `LinearizeOptimallyExample12`
|           31,438.15 |           31,808.49 |    2.4% |      0.01 | `LinearizeOptimallyExample13`
|           57,560.17 |           17,373.13 |    2.1% |      0.01 | `LinearizeOptimallyExample14`
|           17,059.81 |           58,617.31 |    2.3% |      0.01 | `LinearizeOptimallyExample15`
|      193,585,541.00 |                5.17 |    1.3% |      2.15 | `LinearizeOptimallyExample16`
|      196,138,167.00 |                5.10 |    3.0% |      2.24 | `LinearizeOptimallyExample17`
|      182,834,750.00 |                5.47 |    1.8% |      2.07 | `LinearizeOptimallyExample18`
|      200,634,666.00 |                4.98 |    2.2% |      2.20 | `LinearizeOptimallyExample19`

</details>

> That confirms the impression we got: 16-19 seem to just be really hard, and the later optimizations just don't help with it (and increase per-iteration cost, for no meaningful iteration reduction).

16-19 has similar speed with the other clusters in until bb198964e10e84eaa93a31a5b9860bad026e1dc5

> Perhaps it's worth investigating (in a follow up) if there are more specific optimizations targetting the patterns in those clusters.

+1

instagibbs commented:
@ismaelsadeeq I typically overcome that with:
```
  -min-time=<milliseconds>
       Minimum runtime per benchmark, in milliseconds (default: 10)
```

to make it long enough to be stable

instagibbs commented:
reACK a5ea6467b6a694facc1f5efe460bf30539167597

via `git range-diff master 59028de422bbf8ccf53da27f1153e343952e585f a5ea6467b6a694facc1f5efe460bf30539167597`

Just benchmark changes, with examples 11-14 added in 2fcfb4c0392a2f3114195e6e9979c1431798632c and 15-19 in 8fbc6f3c33c6a5b23314279f0856de0625cc6d3f

New benchmarks all run for each commit properly post-introduction.
my benchmark runs for comparison starting from introduction commit to tip: https://gist.github.com/instagibbs/058f51587074187cb78f5f9b9327442e

sdaftuar commented:
I have been gathering more data to try to measure the effect of the improvements in this PR.  I decided to take my cluster mempool branch (#28676), combined with my simulation environment for playing back historical data, and measure the effect of a series of 4 commits from this PR on the number of clusters encountered in 2023 that would NOT have been optimally linearized, for each of 5 different values of the maximum iterations we'd permit when linearizing.  

I think plotted these values on a log-scale, so we can more easily see the effect of the optimizations here. Each data series is a different number of maximum iterations allotted to the linearization algorithm, and I've plotted how the number of clusters that are non-optimally sorted largely decreases with successive commits in this PR:

![Optimal linearizations 2023 by commit and iteration count](https://github.com/user-attachments/assets/954898a9-4caa-40c1-979f-aa23789fd87c)

instagibbs commented:
@sdaftuar is that number of iterations per search invocation, or total over the entire linearization?

Could a version of this graph show the total fraction of *transactions* that are optimally linearized? Or maybe just do a chart where a cluster min size of <larger number> is used to ignore the many small clusters we know we can optimally linearize?

instagibbs reviewed (approved):
reACK 9ad2fe7e69e9e69949ebbb280a15756dc3301f09

just a rebase on master to take up `Revert "build: Minimize I/O operations in `GenerateHeaderFrom{Json,Raw}.cmake`"` I presume

via `git range-diff master a5ea6467b6a694facc1f5efe460bf30539167597 9ad2fe7e69e9e69949ebbb280a15756dc3301f09`

sdaftuar commented:
> @sdaftuar is that number of iterations per search invocation, or total over the entire linearization?

This is the total over the entire linearization, not each search invocation.

> Could a version of this graph show the total fraction of _transactions_ that are optimally linearized? 

This seems a bit difficult to state, since a transaction may be part of many clusters over its lifetime, so I'm not sure precisely what we'd want to measure.

>Or maybe just do a chart where a cluster min size of is used to ignore the many small clusters we know we can optimally linearize? (but I guess if they were easy to linearize, that would be shown in the upper right with 1M?)

Yeah, the idea with the first commit shown is to give a baseline understanding of what fraction of clusters are non-trivial to linearize in the first place.

glozow commented:
My graph on a log scale, looks similar to sipa's:

![image](https://github.com/user-attachments/assets/59a7aebc-f56a-4d1f-add6-f76d5e862c46)

Results txt: [2024-09-14-bench.txt](https://github.com/user-attachments/files/17006092/2024-09-14-bench.txt)

glozow commented on src/bench/cluster_linearize.cpp:
```diff
@@ -242,12 +290,84 @@ static void MergeLinearizations64TxWorstCase(benchmark::Bench& bench) { BenchMer
+// selecting for ones that take many iterations (after the introduction of some but not all
+// linearization algorithm optimizations).
+
+/* 2023-05-05T23:12:21Z 71, 521780, 543141,*/
```
Are these comments in the format timestamp, tx count, fee, size (or similar)? Or are they actually something like iteration count? I added a print statement to get feerates of clusters but got something slightly different.

```
total feerate: 4818977 / 54585, 71 txns
total feerate: 15029636 / 99008, 81 txns
total feerate: 1288344 / 98145, 90 txns
total feerate: 15755611 / 77992, 87 txns
total feerate: 429736 / 8245, 35 txns
total feerate: 695118 / 60056, 60 txns
total feerate: 941605 / 80052, 99 txns
total feerate: 1948779 / 51103, 52 txns
total feerate: 11583092 / 76049, 69 txns
total feerate: 871078 / 55437, 77 txns
total feerate: 6127495 / 94464, 48 txns
total feerate: 1176444 / 14638, 77 txns
total feerate: 76548 / 7724, 40 txns
total feerate: 10487482 / 100707, 96 txns
total feerate: 14575007 / 90755, 93 txns
total feerate: 867760 / 10558, 55 txns
total feerate: 1205280 / 14663, 76 txns
total feerate: 1493360 / 18235, 98 txns
total feerate: 1446604 / 17766, 99 txns
total feerate: 457656 / 23949, 78 txns
```

sdaftuar replied:
The format was timestamp (from my simulation), tx count, min iterations when running Linearize 10 times with different rng seeds and then once more when running it woth LIMO using the optimal linearization as input, and then max iterations over the same set of 11 runs.

glozow replied:
aha, that makes much more sense!

instagibbs replied:
would be worth adding that description if touched again or in followup

ismaelsadeeq replied:
+1 on adding this in a follow-up.
I can create a follow for it see https://github.com/ismaelsadeeq/bitcoin/commit/9350732b17915d315cf3ce9c744d1491c1b5d386 with the description @sdaftuar provided

glozow reviewed (commented):
reACK 9ad2fe7e69e, just have a question about the docs

sdaftuar commented:
ACK 9ad2fe7e69e9e69949ebbb280a15756dc3301f09

ismaelsadeeq commented:
post merge ACK 9ad2fe7e69e9e69949ebbb280a15756dc3301f09
My bench result locally was consistent with what @glozow and sipa has provided.
